icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,463 @@
+        featureList.adoptElement(cloneString(UnicodeString(u"TB2:").append(b2).append(b3).append(b4), status), status);
+        if (U_FAILURE(status)) { return; }
+    }
+    if (isValid(b3) && isValid(b4) && isValid(b5)) {
+        featureList.adoptElement(cloneString(UnicodeString(u"TB3:").append(b3).append(b4).append(b5), status), status);
+        if (U_FAILURE(status)) { return; }
+    }
+    if (isValid(b4) && isValid(b5) && isValid(b6)) {
+        featureList.adoptElement(cloneString(UnicodeString(u"TB4:").append(b4).append(b5).append(b6), status), status);
+        if (U_FAILURE(status)) { return; }
+    }
+
+    int32_t score = negativeSum;
+    for (int32_t j = 0; j < featureList.size(); j++) {
+        UnicodeString* key = (UnicodeString*)featureList.elementAt(j);
+        if(fModel.containsKey(*key)) {
+            score += (2 * fModel.geti(*key));
+        }
+    }
+
+    if (score > 0) {
+        boundary.addElement(index, status);
+        numBreaks++;
+    }
+}
+
+void CjMLBreakEngine::initElementList(const UnicodeString &inString, UVector &elementList,
+                                      UErrorCode &status) const {
+    const UChar32 w1 = INVALID;
+    const UChar32 w2 = INVALID;
+    const UChar32 w3 = inString.char32At(0);
+    const UChar32 w4 = inString.char32At(1);
+    const UChar32 w5 = 2 < inString.length() ? inString.char32At(2) : INVALID;
+    const UChar32 w6 = 3 < inString.length() ? inString.char32At(3) : INVALID;
+
+    const UnicodeString b1(INVALID);
+    const UnicodeString b2(b1);
+    const UnicodeString b3(getUnicodeBlockIndex(w3, status));
+    const UnicodeString b4(getUnicodeBlockIndex(w4, status));
+    const UnicodeString b5(getUnicodeBlockIndex(w5, status));
+    const UnicodeString b6(getUnicodeBlockIndex(w6, status));
+
+    LocalPointer<Element> el1(new Element(w1, b1), status);
+    elementList.adoptElement(el1.orphan(), status);
+    if (U_FAILURE(status)) { return; }
+    LocalPointer<Element> el2(new Element(w2, b2), status);
+    elementList.adoptElement(el2.orphan(), status);
+    if (U_FAILURE(status)) { return; }
+    LocalPointer<Element> el3(new Element(w3, b3), status);
+    elementList.adoptElement(el3.orphan(), status);
+    if (U_FAILURE(status)) { return; }
+    LocalPointer<Element> el4(new Element(w4, b4), status);
+    elementList.adoptElement(el4.orphan(), status);
+    if (U_FAILURE(status)) { return; }
+    LocalPointer<Element> el5(new Element(w5, b5), status);
+    elementList.adoptElement(el5.orphan(), status);
+    if (U_FAILURE(status)) { return; }
+    LocalPointer<Element> el6(new Element(w6, b6), status);
+    elementList.adoptElement(el6.orphan(), status);
+}
+
+UnicodeString CjMLBreakEngine::getUnicodeBlockIndex(UChar32 ch, UErrorCode &status) const{
+    if (U_FAILURE(status)) {
+        return UnicodeString(INVALID);
+    }
+
+    UBlockCode block = ublock_getCode(ch);
+    if (block == UBLOCK_GENERAL_PUNCTUATION ||
```

Comment:
BudouX may need to look for other Unicode blocks in the future as we expand the language coverage.
Is it possible not to hard code the uBlockCode to check?

Reply:
Done.